localstack/extensions/patterns/webapp.py

```diff
@@ -0,0 +1,321 @@
+
+        Here are some examples::
+
+            class MyRoutes:
+                @route("/hello")
+                def hello(request):
+                    return "Hello World!"
+
+            class MyExtension(WebAppExtension):
+                name = "my-extension"
+
+                def collect_routes(self, routes: list[t.Any]):
+
+                    # scans all routes of MyRoutes
+                    routes.append(MyRoutes())
+                    # use rule adapters to add routes without decorators
+                    routes.append(RuleAdapter("/say-hello", self.say_hello))
+
+                    # no idea why you would want to do this, but you can :-)
+                    @route("/empty-dict")
+                    def _inline_handler(request: Request) -> Response:
+                        return Response.for_json({})
+                    routes.append(_inline_handler)
+
+                def say_hello(request: Request):
+                    return {"message": "Hello World!"}
+
+        This creates the following routes available through both subdomain and submount.
+
+        With subdomain:
+
+        * ``my-extension.localhost.localstack.cloud:4566/hello``
+        * ``my-extension.localhost.localstack.cloud:4566/say-hello``
+        * ``my-extension.localhost.localstack.cloud:4566/empty-dict``
+        * ``my-extension.localhost.localstack.cloud:4566/static``  <- automatically added static file endpoint
+
+        With submount:
+
+        * ``localhost.localstack.cloud:4566/_extension/my-extension/hello``
+        * ``localhost.localstack.cloud:4566/_extension/my-extension/say-hello``
+        * ``localhost.localstack.cloud:4566/_extension/my-extension/empty-dict``
+        * ``localhost.localstack.cloud:4566/_extension/my-extension/static``  <- automatically added static file endpoint
+
+        :param routes: the routes being collected
+        """
+        pass
+
+    @cached_property
+    def template_env(self) -> t.Optional["jinja2.Environment"]:
+        """
+        Returns the singleton jinja2 template environment. By default, the environment uses a ``PackageLoader`` that
+        loads from ``my_extension.templates`` (where ``my_extension`` is the root module of the extension,
+        and ``templates`` refers to ``self.template_package_path``, which is ``templates`` by default).
+
+        :return: a template environment
+        """
+        if self.template_package_path:
+            return self._create_template_env()
+        return None
+
+    def _create_template_env(self) -> "jinja2.Environment":
+        """
+        Factory method to create the jinja2 template environment.
+        :return: a new jinja2 environment
+        """
+        import jinja2
+
+        return jinja2.Environment(
+            loader=jinja2.PackageLoader(
+                self.get_extension_module_root(), self.template_package_path
+            ),
+            autoescape=jinja2.select_autoescape(),
```

Comment:
Let's hope this jinja2 API never changes because the peer dependency in the extension needs to package a compatible version 🤞

Reply:
for now it should be OK since we have a pin on jinja through moto. in the worst case people can override the method in their extension.

but the issue of dependencies in extensions (and potential conflicts) is something we definitely need to tackle at some point.